pkg/controller/attach_detach.go

```diff
@@ -152,31 +151,31 @@ func (r *ReconcileAttachDetach) triggerAttach(ctx context.Context, azVolumeAttac
 
 	// initiate goroutine to attach volume
 	go func() {
-		response, err := r.attachVolume(ctx, azVolumeAttachment.Spec.VolumeID, azVolumeAttachment.Spec.NodeName, azVolumeAttachment.Spec.VolumeContext)
+		response, attachErr := r.attachVolume(ctx, azVolumeAttachment.Spec.VolumeID, azVolumeAttachment.Spec.NodeName, azVolumeAttachment.Spec.VolumeContext)
 		var updateFunc func(interface{}) error
-		if err != nil {
-			klog.Errorf("failed to attach volume %s to node %s: %v", azVolumeAttachment.Spec.UnderlyingVolume, azVolumeAttachment.Spec.NodeName, err)
+		if attachErr != nil {
+			klog.Errorf("failed to attach volume %s to node %s: %v", azVolumeAttachment.Spec.UnderlyingVolume, azVolumeAttachment.Spec.NodeName, attachErr)
 
 			updateFunc = func(obj interface{}) error {
 				azv := obj.(*v1alpha1.AzVolumeAttachment)
-				azv = r.updateError(ctx, azv, err)
-				_, err = r.updateState(ctx, azv, v1alpha1.AttachmentFailed)
-				return err
+				azv = r.updateError(ctx, azv, attachErr)
+				_, uerr := r.updateState(ctx, azv, v1alpha1.AttachmentFailed)
+				return uerr
 			}
 		} else {
 			klog.Infof("successfully attached volume (%s) to node (%s) and update status of AzVolumeAttachment (%s)", azVolumeAttachment.Spec.UnderlyingVolume, azVolumeAttachment.Spec.NodeName, azVolumeAttachment.Name)
 
 			updateFunc = func(obj interface{}) error {
 				azv := obj.(*v1alpha1.AzVolumeAttachment)
 				azv = r.updateStatusDetail(ctx, azv, response)
-				_, err := r.updateState(ctx, azv, v1alpha1.Attached)
-				return err
+				_, uerr := r.updateState(ctx, azv, v1alpha1.Attached)
+				return uerr
 			}
 		}
 		if derr := azureutils.UpdateCRIWithRetry(ctx, r.client, r.azVolumeClient, azVolumeAttachment, updateFunc); derr != nil {
-			klog.Errorf("failed to update AzVolumeAttachment (%s) with attachVolume result (response: %v, error: %v): %v", azVolumeAttachment.Name, response, err, derr)
+			klog.Errorf("failed to update AzVolumeAttachment (%s) with attachVolume result (response: %v, error: %v): %v", azVolumeAttachment.Name, response, attachErr, derr)
 		} else {
-			klog.Infof("Successfully updated AzVolumeAttachment (%s) with attachVolume result (response: %v, error: %v)", azVolumeAttachment.Name, response, err)
+			klog.Infof("Successfully updated AzVolumeAttachment (%s) with attachVolume result (response: %v, error: %v)", azVolumeAttachment.Name, response, attachErr)
```

Comment:
The `attachVolume` error was being overwritten by the error returned by `r.updateState()` (OG line 163) in the error case. I renamed the `attachVolume` error and `r.updateState` error variables to improve clarity and prevent accidental overwrites in the future.